docs/contributing/Nullable Annotations.md

```diff
@@ -0,0 +1,67 @@
+# Nullable Annotations
+
+This document describes how nullable annotations should be approached in the 
+Roslyn code base. The default is to simply follow [the same guidance](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/api-guidelines/nullability.md)
+as the [dotnet/runtime](github.com/dotnet/runtime) repository. This document
+serves to detail the places where the guidance differs for Roslyn and 
+re-emphasize rules that come up frequently.
+
+## Annotation Guidance 
+
+- **DO** annotate all new code
+- **DO** use the null suppression operator in places where a method body has
+ensured a value cannot be `null` but the compiler is unable to track it. 
+Example:
+```cs
+struct SyntaxTrivia {
+    void Example() {    
+        if (this.HasStructure) {
+            Use(this.GetStructure()!)
+        }
+    }
+}
+```
+- **DO** use `Assert` to validate state that is passed into the method
```

Comment:
I'm not sure what you mean by "historically" here. Nullable annotations are very new.

I also don't see the reason why we should differ between compiler and IDE here. Can you give reasons why we should differ on this? Is there an IDE style we should be considering instead? 

Having different rules for different layers just makes a hard problem that much harder. Better to standardize and move forward together.

Reply:
```suggestion
- **CONSIDER** using `Assert` to validate state that is passed into the method if it avoids the need for suppressions later in the method
```

Reply:
Historically we didn't use Assert for any null checking, we always did Contract.ThrowIfNull(). This was the policy that predates null annotations and honestly goes back many many years. More or less the logic was the same reason Go doesn't offer asserts: because they're not in retail code, it means you end up looking at Watsons that didn't behave how the developer imagined.

> Having different rules for different layers just makes a hard problem that much harder. Better to standardize and move forward together.

Agreed -- this something we never agreed on historically, this document just reminds us of that. :smile:

Reply:
Ah okay. I have no objection to `Contract.ThrowIfNull`. I was more getting that wee should use a form fo explicit validation, Assert, ThrowIfNull, for external state. Wasn't trying to mandate `Assert` specifically. Let me update that language.

---
In reply to: [366084477](https://github.com/dotnet/roslyn/pull/40940#discussion_r366084477) [](ancestors = 366084477)